src/helpers.c

```diff
@@ -149,130 +149,108 @@ void concat_str_array(char **output, char *const argv[])
+int system_argv_pipe_fd(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+                        int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	pid_t monitorpid = fork();
+	if (monitorpid == -1) {
+		LOG(NULL, "Failed to create child process to monitor pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
+		return -1;
+	} else if (monitorpid == 0) {
+		int pipefd[2];
+		if (pipe(pipefd) == -1) {
+			LOG(NULL, "Failed to create a pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
 			assert(0);
 		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
+		pid_t lhspid = system_argv_fd_nowait(-1, pipefd[1], -1 , lhscmd);
+		if (close(pipefd[1]) == -1) {
+			LOG(NULL, "Could not close write end of pipe file descriptor", "%d", pipefd[1]);
 			assert(0);
 		}
-
-		if (status != 0) {
-			char *cmdline = NULL;
-
-			concat_str_array(&cmdline, argv);
-			LOG(NULL, "Failed to run command:", "%s", cmdline);
-			free(cmdline);
+		pid_t rhspid = system_argv_fd_nowait(pipefd[0], -1, -1, rhscmd);
+		if (close(pipefd[0]) == -1) {
+			LOG(NULL, "Could not close read end of pipe file descriptor", "%d", pipefd[0]);
+			assert(0);
 		}
+
+		int lhsresult = wait_process_terminate(lhspid);
+		int rhsresult = wait_process_terminate(rhspid);
+		exit(rhsresult != EXIT_SUCCESS ? rhsresult : lhsresult);
 	}
+	return wait_process_terminate(monitorpid);
+}
 
-	return status;
+int system_argv(char *const argv[])
+{
+	return system_argv_fd(-1, -1, -1, argv);
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv_fd(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
-		}
-		if (newstdout >= 0) {
-			if (dup2(newstdout, STDOUT_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdout", "");
-				assert(0);
-			}
-			close(newstdout);
-		}
-		if (newstderr >= 0) {
-			if (dup2(newstderr, STDERR_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stderr", "");
-				assert(0);
-			}
-			close(newstderr);
-		}
+	pid_t cmdpid = system_argv_fd_nowait(newstdinfd, newstdoutfd, newstderrfd, cmd);
+	return wait_process_terminate(cmdpid);
+}
 
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
+pid_t system_argv_fd_nowait(int newstdinfd, int newstdoutfd, int newstderrfd, char *const cmd[])
+{
+	pid_t cmdpid = fork();
+	if (cmdpid == -1) {
+		LOG(NULL, "Failed to fork to execute command", "%s", cmd[0]);
 		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+	} else if (cmdpid == 0) {
+		set_std_fd(newstdinfd, newstdoutfd, newstderrfd);
```

Comment:
Roger that.  I'm working on the changes for this, and adding the additional parameter to system_argv_fd_nowait seems to work well.

In my first stab at the second revision for this, before I committed, I had the code symmetric, but I had moved the first close up so that pipe file descriptor was closed a bit earlier.  However, the symmetry makes the code easier to read.  I'll move it back.